# parameter estimates using logistic growth model

In [75]:
from datetime import datetime, timedelta, date
import math 
import pandas as pd 
import matplotlib as mpl 

### TMRCA

In [49]:
youngest_tip = 2022.7972602739726 ## for 528 tips, youngest_tip = 2022.8 

rootHeight_mean = 0.8276
rootHeight_upper = 0.991
rootHeight_lower = 0.6997
std_error_of_mean = 0.0051618

TMRCA_mean = round(youngest_tip, 3) - round(rootHeight_mean, 3)
TMRCA_95CI_upper = round(youngest_tip,3) - round(rootHeight_upper, 3)
TMRCA_95CI_lower = round(youngest_tip,3) - round(rootHeight_lower,3)

def format_date(number):

    year = int(number)
    d = timedelta(days=(number - year)*365)
    day_one = datetime(year,1,1)
    date = d + day_one
    return date.date()



Mean TMRCA: 2021.969 +/- 0.005 or 2021-12-20
95% CI: between 2021-10-22 and 2022-02-05 ( 2021.806 , 2022.097 ) 

December 20, 2021
October 22, 2021
February 05, 2022


In [33]:
print('Mean TMRCA:', TMRCA_mean, '+/-', round(std_error_of_mean,3), 'or', format_date(TMRCA_mean))
print('95% CI:','between', format_date(TMRCA_95CI_upper), 'and', format_date(TMRCA_95CI_lower) ,'(', TMRCA_95CI_upper, ',', TMRCA_95CI_lower, ') \n')


print(format_date(TMRCA_mean).strftime("%B %d, %Y"))
print(format_date(TMRCA_95CI_upper).strftime("%B %d, %Y"))
print(format_date(TMRCA_95CI_lower).strftime("%B %d, %Y"))


og 2022.7972602739726
year 2022
rem 0.7972602739725971
month 8


In [62]:
## parameter estimates from tracer
k = [9.67, 25.3132, 40.2175] 
k_mean_error = 2.5524
r = [8.3098, 14.4011,20.5965]
r_mean_error = 0.154
t50_years = [0.282, 0.3975, 0.5062] 
t50_mean_error = .0036164

[30.445825519793505, 17.56801361731951, 12.283578321772147]

### doubling time

In [64]:
d = []
for i in range(len(r)): 
    d.append(round((math.log(2)/r[i])*365,3)) 

d

[30.446, 17.568, 12.284]

### exponential gowth rate

In [65]:
r_days = []
for i in range(len(r)): 
    r_days.append(round(r[i]/365,3))

r_days

[0.023, 0.039, 0.056]

### fundamental reproductive number

SIR dynamics: 
per-day rate of increase (r) = contact rate (β) - recovery rate (γ) 

solve for β: r + γ 

find R0: R0 = β / γ


In [70]:
# 21 day duration of infection (14-28) // WHO monkeypox fact sheet

gamma = round(1/21, 3) # recovery rate per day 

r_mean = r_days[1]

beta = r_mean + gamma

r0 = round(beta/gamma, 3) 


print('γ =',gamma)
print('r =', r_mean)
print('β =',beta)
print('R0 =',r0)

γ = 0.048
r = 0.039
β = 0.087
R0 = 1.812


### Early generation time

τ = 1 / (2 β) 

Accoriding to WHO, incubation period is 5-21 days; 
early in the pandemic, S = 1, so generation time only depends on contact rate -> 5.747 days 

In [74]:
tau_early = 1/(2*beta)

print('τ (days) =', round(tau_early,3))
print('τ (years) =', round(tau_early/365,3))  

τ (days) = 5.747
τ (years) = 0.016


 ### 
 


## Netau through time

### read in in log file

In [76]:
t50_file = "/Users/nashwa/Desktop/git/hmpxv_dynamics/out/logistic/logistic_growth_t50"

t50_table = pd.read_csv(t50_file,sep='\t')

t50s = t50_table[] 

Unnamed: 0  \
time               date       datetime              milliseconds   
2022.7972602739726 2022-10-18 2022-10-18T08:00:00  1666080000846   
2022.7901922540573 2022-10-15 2022-10-15T18:05:03  1665857103770   
2022.783124234142  2022-10-13 2022-10-13T04:10:06  1665634206694   
2022.7760562142266 2022-10-10 2022-10-10T14:15:09  1665411309618   
...                                                          ...   
2022.125798382017  2022-02-15 2022-02-15T05:59:38  1644904778628   
2022.1187303621016 2022-02-12 2022-02-12T16:04:41  1644681881552   
2022.1116623421863 2022-02-10 2022-02-10T02:09:44  1644458984476   
2022.104594322271  2022-02-07 2022-02-07T12:14:47  1644236087400   
2022.0975263023556 2022-02-04 2022-02-04T22:19:50  1644013190324   

                                                  Logistic t50: 235hmpxv_logistic.log  \
time               date       datetime                                           mean   
2022.7972602739726 2022-10-18 2022-10-18T08:00:00                  25.313161588084533   
2022.7901922540573 2022-10-15 2022-10-15T18:05:03                   24.82677653408336   
2022.783124234142  2022-10-13 2022-10-13T04:10:06                    24.4567159808162   
2022.7760562142266 2022-10-10 2022-10-10T14:15:09                  24.156761965840328   
...                                                                               ...   
2022.125798382017  2022-02-15 2022-02-15T05:59:38                  0.5811366768472955   
2022.1187303621016 2022-02-12 2022-02-12T16:04:41                   0.532593274662271   
2022.1116623421863 2022-02-10 2022-02-10T02:09:44                  0.4881724609159278   
2022.104594322271  2022-02-07 2022-02-07T12:14:47                  0.4475270471155334   
2022.0975263023556 2022-02-04 2022-02-04T22:19:50                 0.41033736801709175   

                                                            Unnamed: 2  \
time               date       datetime                          median   
2022.7972602739726 2022-10-18 2022-10-18T08:00:00   20.311422837120407   
2022.7901922540573 2022-10-15 2022-10-15T18:05:03    20.29793931849625   
2022.783124234142  2022-10-13 2022-10-13T04:10:06    20.27838927343065   
2022.7760562142266 2022-10-10 2022-10-10T14:15:09   20.249127208538543   
...                                                                ...   
2022.125798382017  2022-02-15 2022-02-15T05:59:38   0.4560852452062752   
2022.1187303621016 2022-02-12 2022-02-12T16:04:41   0.4144682837157928   
2022.1116623421863 2022-02-10 2022-02-10T02:09:44  0.37447511690437973   
2022.104594322271  2022-02-07 2022-02-07T12:14:47   0.3392584185752622   
2022.0975263023556 2022-02-04 2022-02-04T22:19:50  0.30734688421626544   

                                                           Unnamed: 3  \
time               date       datetime                          upper   
2022.7972602739726 2022-10-18 2022-10-18T08:00:00   49.88837554388796   
2022.7901922540573 2022-10-15 2022-10-15T18:05:03   49.80447725367827   
2022.783124234142  2022-10-13 2022-10-13T04:10:06  49.491179441456346   
2022.7760562142266 2022-10-10 2022-10-10T14:15:09   49.10957347243773   
...                                                               ...   
2022.125798382017  2022-02-15 2022-02-15T05:59:38  1.8019867574506072   
2022.1187303621016 2022-02-12 2022-02-12T16:04:41  1.7104604147697249   
2022.1116623421863 2022-02-10 2022-02-10T02:09:44   1.583785750866055   
2022.104594322271  2022-02-07 2022-02-07T12:14:47  1.4810351949694633   
2022.0975263023556 2022-02-04 2022-02-04T22:19:50  1.3821980607353848   

                                                             Unnamed: 4  
time               date       datetime                            lower  
2022.7972602739726 2022-10-18 2022-10-18T08:00:00    11.432421391925663  
2022.7901922540573 2022-10-15 2022-10-15T18:05:03    11.431847650959531  
2022.783124234142  2022-10-13 2022-10-13T04:10:06    11.431197292235618  
2022.7760562142266 2022-10-10 20

In [1]:
#new df: t50s: 95 CI upper, mean,  95 CI lower, TIME, DATE 
k[1] 
r_mean 
tau_early 


NameError: name 'k' is not defined

In [ ]:
# 13 day incubation period (5-21) // WHO monkeypox fact sheet
# tau = 13

In [ ]:
## notes

## https://github.com/trvrb/dynamics-practical/blob/master/examine-the-logistic-growth-output.md

## carrying capacity k 

 

## doubling time t50 
# Ne * τ

t50 = log(2) / r * 365 #doubling time in years
#use doubling time in days to quantify the exponential growth rate

## growth rate Rt

In [ ]:
# run parameter logistic growth + do parametric plotting, 
# e.g. file:///Users/nashwa/Downloads/ncov-wa-clade.pdf

Population size =
1 / rate of coalescence =
Ne * τ = 

Ne * (1 / (2 * β) ) = 

Ne * (1 / (2 * 0.087) ) = 

logistic model estimates Netau -> derive Ne at each timepoint; 


** should tau change over time? estimated as 'early epidemic rate' when S = 1; shifts over course of year ->  